src/security/Handshake.cc

```diff
@@ -504,6 +506,50 @@ Security::HandshakeParser::parseSniExtension(const SBuf &extensionData) const
     return SBuf(); // SNI extension lacks host_name
 }
 
+// RFC 8446 Section 4.2.1: SupportedVersions extension
+void
+Security::HandshakeParser::parseSupportedVersionsExtension(const SBuf &extensionData) const
+{
+    // RFC 8446 Section 4.1.2:
+    // In TLS 1.3, the client indicates its version preferences in the
+    //  "supported_versions" extension (Section 4.2.1) and the
+    // legacy_version field MUST be set to 0x0303, which is the version
+    // number for TLS 1.2.
+    //
+    // RFC 8446 Section 4.2.1:
+    // A server which negotiates TLS 1.3 MUST respond by sending
+    // a "supported_versions" extension containing the selected version
+    // value (0x0304).  It MUST set the ServerHello.legacy_version field to
+    // 0x0303 (TLS 1.2).
+    //
+    // For both Client and Server hello the legacy_version (highest supported
+    // version for TLSv1.2 and earlier) must be set to 0x0303
+    if (details->tlsSupportedVersion != AnyP::ProtocolVersion(AnyP::PROTO_TLS, 1, 3))
```

Comment:
Are we trying to check the (initial) legacy version here? Or the finalized version? Or both, depending on how many versions extensions we receive?The comments above talk about TLS v1.2 but this checks TLS v1.3. 

Overall, this is still very confusing. You have described what the protocol requires the agents to do. That part is OK. Can you now describe (via a GitHub comment?) what our Squid code is trying to do? Once I understand what is going on, I can try to integrate everything.

Reply:
> Are we trying to check the (initial) legacy version here? Or the finalized version? Or both, depending on how many versions extensions we receive?The comments above talk about TLS v1.2 but this checks TLS v1.3.

There is a bug here added with commit 98e8600.
We must check here for TLSv1.2

My sense is that this is still confused mostly because of naming scheme we selected (tlsSupportedVersion) which conflicts with the supported_version extension and the legacy_version naming scheme which if I am not wrong added with the TLSv1.3 RFC.
Moreover the TLSv1.3 is confusing trying to be compatible with older TLS releases. 

> What our Squid code is trying to do?

It tries to detect TLSv.1.3 hello messages. It is a TLSv1.3 if the tlsSupportedVersion is set to TLSv1.2 and the "supported_versions" extension containing the TLSv1.3 value.
It uses this trick to make the TLS hello to looks like a TLSv1.2 hello message.